### DISCUSSION TITLE
Help with Understanding Print Times

### DISCUSSION THREAD

**Discussion by joehhn** [sn4k3/UVtools#658] [General]
Hello! 

How does UVTools 3.11.2 calculate its approximate print times?In my own math, I consistently underestimate print times by about 20%, but UVTools usually hits +/- 1 min from the actual time. Currently accounting for wait-before-cure times, exposure for normal/trans/bottom layers, and per-layer lift/retracts. Not sure what else could be besides the first retract/final lift, but I'd love to understand how UVTools determines these under the hood. Thanks :)

**Comment by sn4k3** [comment]
UVtools sums all instant times which is not accurate by far depending on your machine acceleration and layer count. To understand what is under the hood the best is to read the code that does that calculation: 

https://github.com/sn4k3/UVtools/blob/31416b619a08914b50cc8ef28222c3795a2e7476/UVtools.Core/FileFormats/FileFormat.cs#L2812-L2888

**Comment by joehhn** [comment]
Just what I needed, thank you! One quick question: can you tell me whether these calculations assume each normal layer has the same constant exposure time "layer.ExposureTime" (i.e., no support for different exposures as set in "Dynamic Layer Heights")? Or is ExposureTime an instance variable that changes depending on the specific layer? Don't know much at all about C#. Thanks again :)

**Reply by sn4k3** [reply]
It sums data from layers, if layers have different time values it will take them into account.

**Reply by joehhn** [reply]
Great! Thanks for clearing that up for me.
